pkg/traceql/ast.go

```diff
@@ -475,204 +476,406 @@ func (o UnaryOperation) referencesSpan() bool {
+	return Static{
+		Type:      TypeDuration,
+		valScalar: uint64(d),
+	}
 }
 
-func (s Static) compare(other *Static) int {
-	if s.Type != other.Type {
-		if s.asFloat() > other.asFloat() {
-			return 1
-		} else if s.asFloat() < other.asFloat() {
-			return -1
-		}
+func NewStaticStatus(s Status) Static {
+	return Static{
+		Type:      TypeStatus,
+		valScalar: uint64(s),
+	}
+}
 
-		return 0
+func NewStaticKind(k Kind) Static {
+	return Static{
+		Type:      TypeKind,
+		valScalar: uint64(k),
 	}
+}
 
+func NewStaticIntArray(i []int) Static {
+	if i == nil {
+		return Static{Type: TypeIntArray}
+	}
+	if len(i) == 0 {
+		return Static{Type: TypeIntArray, valBytes: []byte{}}
+	}
+
+	numBytes := uintptr(len(i)) * unsafe.Sizeof(i[0])
+	return Static{
+		Type:     TypeIntArray,
+		valBytes: unsafe.Slice((*byte)(unsafe.Pointer(&i[0])), numBytes),
+	}
+}
+
+func NewStaticFloatArray(f []float64) Static {
+	if f == nil {
+		return Static{Type: TypeFloatArray}
+	}
+	if len(f) == 0 {
+		return Static{Type: TypeFloatArray, valBytes: []byte{}}
+	}
+
+	numBytes := uintptr(len(f)) * unsafe.Sizeof(f[0])
+	return Static{
+		Type:     TypeFloatArray,
+		valBytes: unsafe.Slice((*byte)(unsafe.Pointer(&f[0])), numBytes),
+	}
+}
+
+func NewStaticStringArray(s []string) Static {
+	if s == nil {
+		return Static{Type: TypeStringArray}
+	}
+	if len(s) == 0 {
+		return Static{Type: TypeStringArray, valStrings: []string{}}
+	}
+
+	return Static{
+		Type:       TypeStringArray,
+		valStrings: s,
+	}
+}
+
+func NewStaticBooleanArray(b []bool) Static {
+	if b == nil {
+		return Static{Type: TypeBooleanArray}
+	}
+	if len(b) == 0 {
+		return Static{Type: TypeBooleanArray, valBytes: []byte{}}
+	}
+
+	return Static{
+		Type:     TypeBooleanArray,
+		valBytes: unsafe.Slice((*byte)(unsafe.Pointer(&b[0])), len(b)),
+	}
+}
+
+var seedBytes = []byte{204, 38, 247, 160, 15, 37, 67, 77}
+
+func (s Static) MapKey() StaticMapKey {
 	switch s.Type {
-	case TypeInt:
-		if s.N > other.N {
-			return 1
-		} else if s.N < other.N {
-			return -1
+	case TypeNil:
+		return StaticMapKey{typ: TypeNil}
+	case TypeString:
+		var str string
+		if len(s.valBytes) > 0 {
+			str = unsafe.String(unsafe.SliceData(s.valBytes), len(s.valBytes))
 		}
-	case TypeFloat:
-		if s.F > other.F {
-			return 1
-		} else if s.F < other.F {
-			return -1
+		return StaticMapKey{typ: s.Type, str: str}
+	case TypeIntArray, TypeFloatArray, TypeBooleanArray:
+		if len(s.valBytes) == 0 {
+			return StaticMapKey{typ: s.Type}
 		}
-	case TypeDuration:
-		if s.D > other.D {
-			return 1
-		} else if s.D < other.D {
-			return -1
+
+		sum := crc32.ChecksumIEEE(s.valBytes)
+		return StaticMapKey{typ: s.Type, code: uint64(sum)}
+	case TypeStringArray:
+		if len(s.valStrings) == 0 {
+			return StaticMapKey{typ: s.Type}
 		}
-	case TypeString:
-		if s.S > other.S {
-			return 1
-		} else if s.S < other.S {
-			return -1
+
+		crc := crc32.NewIEEE()
+		_, _ = crc.Write(seedBytes) // avoid collisions with values like []string{""}
```

Comment:
use fnv1a? i believe it's faster than a crc. i know it's not the best, but we use it pretty exclusively in tempo

Reply:
I think it makes sense to use the same hash function consistently throughout Tempo

On amd64 hardware where sse and crc32 instructions are available, crc32 is quite a bit faster than fnv. But I don't think this is true for the 64bit variants and crc and fnv perform more or less the same.

I'll change it to fnv1a 64 bit, but it could be an interesting hackweek project to look into other algorithms

Reply:
nice! my impression is that even fnv1a is considered slow by modern standards. i've always wanted to experiment with other options.